### PR TITLE
Various Improvements

### PR DESCRIPTION
**Pull request by CalvinWilkinson**
<h1 style="font-weight:bold" align="center">Preview Feature Pull Request</h1>
<h2 style="font-weight:bold" align="center">✅Code Review Checklist✅</h2>

- [x] The **_head(source)_** branch for this pull request is a **_preview feature_** branch, with the correct naming syntax, in the following example:
  - A **_head(source)_** branch is the branch being merged into a target branch.
    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/feature/<issue-num>-<description>
      Example: preview/feature/123-my-preview-feature
      ```
    </details>

- [x] The **_head(source)_** branch for this pull request is created from a **_preview release_** branch, with the correct naming syntax, in the following example:
  <details closed><summary>Example</summary>

    ``` xml
    Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
    Example: preview/v1.2.3-preview.4
    ```
  </details>

- [x] The **_base(target)_** branch for this pull request is a **_preview release_** branch, with the correct naming syntax, in the following example:
  - A **_base(target)_** branch is the branch that the **_head(source)_** branch is merging into.

    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
      Example: preview/v1.2.3-preview.4
      ```
    </details>

💡For more information on branching, refer to the project's [branching documentation](../../Documentation/Branching.md).

- [x] Associated issue exists and is linked to this pull request.
  - One issue per pull request.

- [x] Pull request title matches the title of the associated issue.

- [x] A preview-label label has been added to the pull request.

- [x] The labels attached to this PR match the labels attached to the associated issue.

- [x] My code follows the coding style of this project.
  - The style is enforced by the **_.editorconfig_** files in the project and displayed as warnings.  If there is an edge case, reach out and let's discuss it.

- [x] I have manually tested my code changes to the best of my ability.

- [x] All tests passed locally.
  - This is required because unit tests might pass locally but not in the CI environment during the status check process or vice-versa.
  - Tests might pass on the developer's machine but not necessarily on the code reviewer's machine.
  - Status checks run for this pull request when merging into the **_preview/v\*.\*.\*-preview.\*_** branch.  These status checks run every time a change is pushed to the pull request.  These checks validate version syntax, tagging, builds, unit tests, and more.

### CODE REVIEW

File: GitHubData/Exceptions/RateLimitException.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+// <copyright file="RateLimitException.cs" company="KinsonDigital">
+// Copyright (c) KinsonDigital. All rights reserved.
+// </copyright>
+
+namespace GitHubData.Exceptions;
+
+// TODO: Create tests
```
Comment (KinsonDigitalAdmin) [inline GitHubData/Exceptions/RateLimitException.cs:7]: Remove TODO item

File: GitHubData/Exceptions/RateLimitException.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+/// <summary>
+/// Occurs the GitHub request rate limit has been reached.
```
Comment (KinsonDigitalAdmin) [inline GitHubData/Exceptions/RateLimitException.cs:10]: Refactor code docs to below:

```suggestion
/// Occurs when the GitHub request rate limit has been reached.
```

File: GitHubData/Factories/HTTPClientFactory.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+﻿// <copyright file="HTTPClientFactory.cs" company="KinsonDigital">
+// Copyright (c) KinsonDigital. All rights reserved.
+// </copyright>
+
+namespace GitHubData.Factories;
+
+/// <summary>
+/// Creates an <see cref="IHttpClient"/> for making HTTP request.
```
Comment (KinsonDigitalAdmin) [inline GitHubData/Factories/HTTPClientFactory.cs:8]: Refactor code docs to below:

```suggestion
/// Creates an <see cref="IHttpClient"/> for making HTTP requests.
```

File: GitHubData/Factories/HTTPClientFactory.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    private static IHttpClient? httpClientSingleton;
+
+    /// <summary>
+    /// Creates an new <see cref="IHttpClient"/> instance or returns the same <see cref="IHttpClient"/> singleton.
+    /// </summary>
+    /// <param name="useAsSingleton">If <c>true</c>, will return the same instance every time.</param>
+    /// <returns>
+    ///     A new or singleton <see cref="IHttpClient"/> instance.
+    /// </returns>
+    public static IHttpClient CreateHTTPClient(bool useAsSingleton = false)
```
Comment (KinsonDigitalAdmin) [inline GitHubData/Factories/HTTPClientFactory.cs:21]: Refactor name of argument to `createAsSingleton`

```suggestion
    public static IHttpClient CreateHTTPClient(bool createAsSingleton = false)
```

File: GitHubData/Models/CoreModel.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
+// <copyright file="CoreModel.cs" company="KinsonDigital">
+// Copyright (c) KinsonDigital. All rights reserved.
+// </copyright>
+
+namespace GitHubData.Models;
+
+/// <summary>
+/// Core information about the rate limits of an authorized
+/// or unauthorized request.
```
Comment (KinsonDigitalAdmin) [inline GitHubData/Models/CoreModel.cs:9]: Refactor code docs to below:

```suggestion
/// Core information about the rate limits of an authorized or unauthorized request.
```

File: GitHubData/Models/ResourcesModel.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+// <copyright file="ResourcesModel.cs" company="KinsonDigital">
+// Copyright (c) KinsonDigital. All rights reserved.
+// </copyright>
+
+namespace GitHubData.Models;
+
+/// <summary>
+/// Holds various resources information about the rate limits of
+/// an authorized or unauthorized account.
```
Comment (KinsonDigitalAdmin) [inline GitHubData/Models/ResourcesModel.cs:9]: Refactor code docs to below:

```suggestion
/// Holds information about the rate limits of an authorized or unauthorized request.
```

File: GitHubData/Models/ResourcesModel.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+/// </summary>
+public record ResourcesModel
+{
+    /// <summary>
+    /// Gets or sets the core limit info.
```
Comment (KinsonDigitalAdmin) [inline GitHubData/Models/ResourcesModel.cs:14]: Refactor to below:

```suggestion
    /// Gets or sets the core data.
```

File: GitHubData/ResourcesModelJSONConverter.cs [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+// </copyright>
+
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using GitHubData.Models;
+
+namespace GitHubData;
+
+/// <summary>
+/// Converts a JSON to a <see cref="ResourcesModel"/> object.
```
Comment (KinsonDigitalAdmin) [inline GitHubData/ResourcesModelJSONConverter.cs:12]: Refactor code docs to below:

```suggestion
/// Converts JSON data to a <see cref="ResourcesModel"/> object.
```

File: GitHubData/Services/IGitHubDataService.cs [hunk 9]
```diff
@@ -2,13 +2,19 @@
...
-namespace GitHubData;
+namespace GitHubData.Services;
 
 /// <summary>
 /// Provides access to GitHub data.
 /// </summary>
 public interface IGitHubDataService : IDisposable
 {
+    /// <summary>
+    /// Gets or sets the auth token to use when making requests for GitHub data.
```
Comment (KinsonDigitalAdmin) [inline GitHubData/Services/IGitHubDataService.cs:13]: Refactor code docs to below:

```suggestion
    /// Gets or sets the authorization token to use when making requests for GitHub data.
```

File: GitHubData/Services/IGitHubDataService.cs [hunk 10]
```diff
@@ -2,13 +2,19 @@
...
+    /// </summary>
+    /// <remarks>If the token is null or empty, then non-authorized requests will be made.</remarks>
```
Comment (KinsonDigitalAdmin) [inline GitHubData/Services/IGitHubDataService.cs:15]: Refactor code docs to below:

```suggestion
    /// <remarks>If the token is null or empty, then non-authorized requests will be used.</remarks>
```

File: GitHubData/Services/IRequestRateLimitService.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+namespace GitHubData.Services;
+
+/// <summary>
+/// Provides the ability to get request rate limit data for
+/// non-authorized and authorized requests to the GitHub API.
+/// </summary>
+public interface IRequestRateLimitService
+{
+    /// <summary>
+    /// Gets or sets the auth token to use when making requests for GitHub data.
```
Comment (KinsonDigitalAdmin) [inline GitHubData/Services/IRequestRateLimitService.cs:14]: Refactor code docs to below:

```suggestion
    /// Gets or sets the authorization token to use when making requests for GitHub data.
```

File: GitHubData/Services/IRequestRateLimitService.cs [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    /// </summary>
+    /// <remarks>If the token is null or empty, then non-authorized requests will be made.</remarks>
```
Comment (KinsonDigitalAdmin) [inline GitHubData/Services/IRequestRateLimitService.cs:16]: Refactor code docs to below:

```suggestion
    /// <remarks>If the token is null or empty, then non-authorized requests will be used.</remarks>
```

**Review by KinsonDigitalAdmin — CHANGES_REQUESTED** [review]

**Review by KinsonDigitalAdmin — APPROVED** [review]

File: GitHubData/Models/ResourcesModel.cs [hunk 13]
```diff
@@ -0,0 +1,16 @@
+// <copyright file="ResourcesModel.cs" company="KinsonDigital">
+// Copyright (c) KinsonDigital. All rights reserved.
+// </copyright>
+
+namespace GitHubData.Models;
+
+/// <summary>
+/// Holds various resources information about the rate limits of an authorized or unauthorized account.
```
Comment (kselena) [inline GitHubData/Models/ResourcesModel.cs:8]: Grammar
```suggestion
/// Holds various resources' information about the rate limits of an authorized or unauthorized account.
```

File: GitHubData/RateLimit.cs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+/// </summary>
+public class RateLimit
+{
+    /// <summary>
+    /// Gets or sets the limit of the request.
+    /// </summary>
+    public int RequestLimit { get; set; }
+
+    /// <summary>
+    /// Gets or sets the remaining requests available that is allowed
```
Comment (kselena) [inline GitHubData/RateLimit.cs:18]: Grammar
```suggestion
    /// Gets or sets the remaining requests available that are allowed
```

File: GitHubData/ResourcesModelJSONConverter.cs [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+namespace GitHubData;
+
+/// <summary>
+/// Converts JSON data to a <see cref="ResourcesModel"/> object.
+/// </summary>
+public class ResourcesModelJSONConverter : JsonConverter<ResourcesModel>
+{
+    /// <summary>
+    /// Read and convert the JSON to <see cref="ResourcesModel"/>.
```
Comment (kselena) [inline GitHubData/ResourcesModelJSONConverter.cs:17]: Grammar
```suggestion
    /// Reads and converts the JSON to <see cref="ResourcesModel"/>.
```

File: GitHubData/ResourcesModelJSONConverter.cs [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+    /// </summary>
+    /// <remarks>
+    /// A converter may throw any Exception, but should throw <cref>JsonException</cref> when the JSON is invalid.
```
Comment (kselena) [inline GitHubData/ResourcesModelJSONConverter.cs:20]: Grammar
```suggestion
    /// A converter may throw any exception but should throw <cref>JsonException</cref> when the JSON is invalid.
```

File: GitHubData/ResourcesModelJSONConverter.cs [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+    /// </remarks>
+    /// <param name="reader">The <see cref="Utf8JsonReader"/> to read from.</param>
```
Comment (kselena) [inline GitHubData/ResourcesModelJSONConverter.cs:22]: Grammar
```suggestion
    /// <param name="reader">The <see cref="Utf8JsonReader"/> to read.</param>
```

File: GitHubData/ResourcesModelJSONConverter.cs [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+        }
+
+        throw new JsonException();
+    }
+
+    /// <summary>
+    /// Write the value as JSON.
+    /// </summary>
+    /// <remarks>
+    /// A converter may throw any Exception, but should throw <cref>JsonException</cref> when the JSON
```
Comment (kselena) [inline GitHubData/ResourcesModelJSONConverter.cs:86]: Grammar
```suggestion
    /// A converter may throw any exception but should throw <cref>JsonException</cref> when the JSON
```

File: GitHubData/ResourcesModelJSONConverter.cs [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+    /// cannot be created.
+    /// </remarks>
+    /// <param name="writer">The <see cref="Utf8JsonWriter"/> to write to.</param>
```
Comment (kselena) [inline GitHubData/ResourcesModelJSONConverter.cs:89]: Grammar
```suggestion
    /// <param name="writer">The <see cref="Utf8JsonWriter"/> to write.</param>
```

**Review by kselena — CHANGES_REQUESTED** [review]

**Review by kselena — APPROVED** [review]
